### DISCUSSION TITLE
Dashboard-as-Code

### DISCUSSION THREAD

**Discussion by AntoineThebaud** [perses/perses#1727] [Design Discussions]
Just creating this discussion about Dashboard-as-Code, a topic that we already started to work on but whose outlines are not 100% clear yet & that is likely to evolve over time.

**Comment by AntoineThebaud** [comment]
After a live discussion with @celian-garcia @Gladorme @Nexucis, these are the things we have identified & agreed on around the Continuous Integration use case with Dashboard-as-Code (aka the `percli dac ci` command, yet to be implemented):
- It should generate preview dashboards.
- Preview dashboards should be automatically removed (expire) after some time.
- To achieve the latter we plan to add a new resource type, `EphemeralDashboard`, that would basically look like the following:
  ```json
  {
    "kind": "EphemeralDashboard",
    "metadata": {
      "name": "NodeExporter",
      "createdAt": "2023-11-02T10:38:15.6762124Z",
      "updatedAt": "2024-01-24T15:01:37.4571497Z",
      "project": "perses"
    },
    "spec": {
      "ttl": "7d",
      "dashboardSpec": {}
    }
  }
  ```
  We make it a new resource type instead of modifying the current Dashboard, so that it's easy to not mix them in search results / GET calls in general.
- It should leverage on the sub-folder feature https://github.com/perses/perses/issues/183 once available: for a given change (a) dedicated sub-folder(s) should be generated by the CI:
  - A sub-folder in this case should have a prefix indicating the change it's related to (e.g PR number).
  - This may imply the addition of an `EphemeralFolder` resource too, to not have empty folders lasting once all the ephemeral dashboards contained got removed. Point to be further discussed.
- We don't plan to support an integration with git providers like GitHub, GitLab, Bitbucket for now, regarding appending information in pull-requests. We consider this is too specific/opinionated & would lead to too much maintenance. Instead we plan to produce a generic "card" in output of `percli dac ci` containing all the info (list of changes, action done for each (create/update/delete), links to the ephemeral dashboards created, etc), that 3rd-party tools could consume to append the information in the corresponding pull-request.

Out-of-topic notes:
- Ephemeral dashboards could be useful in other cases, like ongoing incidents: we can imagine users could fork a given dashboard to create a temporary copy with e.g variables & timerange saved with the relevant values for this incident.

**Reply by celian-garcia** [reply]
Ah yes something to not forget that we talked about: add the ephemeral dashboard in the /proxy endpoint. In case a datasource resides in the dashboard itself.

**Reply by nicolastakashi** [reply]
Very nice idea folks!
Quick question:
> We make it a new resource type instead of modifying the current Dashboard, so that it's easy to not mix them in search results / GET calls in general.

This means ephemeral dashboards will not be present on the dashboard list for searching?

**Reply by Nexucis** [reply]
yup exactly. It's to avoid to be polluted with a tons of dashboard changes. I don't think you want to see a dashboard coming from a PR when you are in an incident.

**Reply by jgbernalp** [reply]
Nice idea, do you think this could be extended for Dashboards as CRDs?, or even a more general approach having dashboards that can decide where they are stored/fetched, like database (MySQL or any supported provider), any ephemeral store (e.g. Redis), CRDs coming from k8s, from an AP, etc?

**Reply by nicolastakashi** [reply]
> Yup exactly. It's to avoid to be polluted with a tons of dashboard changes. I don't think you want to see a dashboard coming from a PR when you are in an incident.

Cool, I'm just curious how we can find these dashboards afterward.
I mean if I'm a Perse's instance and I want to see the dashboard is being changed, where should I go?

**Reply by sjcobb** [reply]
Did we decide if we’re ever going to supports tags? Seems like having an automated tag applied to these dashboards that gets filtered out by default could be an option, and then if you do want to see them all in the listing, filter by that common tag

**Reply by AntoineThebaud** [reply]
I think we didn't discuss tags yet, I'm not sure it would change much for the filtering here compared to filter based on the presence/value of a new `ttl` field in the dashboard.
Basically we think that in most cases, users wouldn't like to mix regular dashboards with temporary ones, hence this technical choice to go with 2 different resources to optimize that scenario we bet is gonna be the standard usage. But then it doesn't prevent to later implement a mean on the UI to merge the 2 in a single view.

**Comment by Nexucis** [comment]
I think with:

* Ephemeral Dashboard
* GO and Cuelang SDK
* CI workflow
* and the demo done at KubeCon
* perses-operator

we have demonstrated that the dashboard as code feature is completely implemented in Perses, and there is nothing else to discuss.
Other evolution of the current workflow will pass through issue or a different discussion if it requires more thought.
